src/protocols/musig2/mod.rs

```diff
@@ -0,0 +1,167 @@
+#![allow(non_snake_case)]
+
+//! Simple ed25519
+//!
+//! See https://tools.ietf.org/html/rfc8032
+//! This is an implementation of the Musig2 protocol as shown in https://eprint.iacr.org/2020/1261.pdf with the addition named Musig2* suggested in Section B of the paper.
+//! We implement the v = 2 (NUMBER_OF_NONCES) version, meaning there are 2 nonces generated by each party.
+
+use super::{ExpandedKeyPair, Signature};
+use curv::cryptographic_primitives::hashing::DigestExt;
+use curv::elliptic::curves::{Ed25519, Point, Scalar};
+use rand::Rng;
+use sha2::{digest::Digest, Sha512};
+
+const NUMBER_OF_NONCES: usize = 2;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PublicKeyAgg {
+    pub agg_public_key: Point<Ed25519>,
+    pub musig_coefficient: Scalar<Ed25519>,
+}
+
+impl PublicKeyAgg {
+    #[allow(clippy::unnecessary_sort_by)]
+    pub fn key_aggregation_n(
+        mut public_keys: Vec<Point<Ed25519>>,
+        my_public_key: &Point<Ed25519>,
+    ) -> PublicKeyAgg {
+        let mut my_coeff = Scalar::zero();
+        let mut sum = Point::zero();
+        public_keys.sort_by(|left, right| left.to_bytes(false).cmp(&right.to_bytes(false)));
+        let mut second_public_key = &public_keys[0];
+        for public_key in &public_keys[1..] {
+            if public_key
+                .to_bytes(false)
+                .gt(&public_keys[0].to_bytes(false))
+            {
+                second_public_key = public_key;
+                break;
+            }
+        }
+
+        public_keys.iter().for_each(|public_key| {
+            let mut musig_coefficient: Scalar<Ed25519> = Scalar::from(1);
+            if public_key != second_public_key {
+                let mut hasher = Sha512::new().chain(&[1]).chain(&*public_key.to_bytes(true));
+                for pk in &public_keys {
+                    hasher.update(&*pk.to_bytes(true));
+                }
+                musig_coefficient = hasher.result_scalar();
+            }
+
+            let a_i = public_key * &musig_coefficient;
+            if public_key == my_public_key {
+                my_coeff = musig_coefficient;
+            }
+            sum = &sum + a_i
+        });
+        PublicKeyAgg {
+            agg_public_key: sum,
+            musig_coefficient: my_coeff,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct PartialNonces {
+    pub r: [Scalar<Ed25519>; NUMBER_OF_NONCES],
+    pub R: [Point<Ed25519>; NUMBER_OF_NONCES],
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct PartialSignature {
+    pub R: Point<Ed25519>,
+    pub my_partial_s: Scalar<Ed25519>,
+}
+
+pub fn generate_partial_nonces(
+    keys: &ExpandedKeyPair,
+    message: Option<&[u8]>,
+    rng: &mut impl Rng,
+) -> PartialNonces {
```

Comment:
I think the variant with the `Rng` should be private and we should expose a wrapper that passes `thread_rng()` into it. to prevent users from passing insecure randomness (another option is that the wrapper will require a `CryptoRng`, See: https://docs.rs/rand/latest/rand/trait.CryptoRng.html)